### ISSUE TITLE
Cargo workspace dependencies still don't work due to an outdated version of cargo_toml

### ISSUE THREAD

**Issue by joell** [bazelbuild/rules_rust#1771]
TL;DR: Issue #1574 is still a problem but can probably easily be resolved by upgrading `crate_universe`'s `cargo_toml` dependency from 0.13 to 0.14.

## Summary

Cargo's [`workspace.dependencies`](https://doc.rust-lang.org/stable/cargo/reference/workspaces.html#the-dependencies-table) feature still causes breakage during repinning/updating of dependencies.

## Environment
Bazel 6.0.0
Rust 1.66.1
rules_rust 0.16.1

## How to reproduce
_Copied from issue #1574 and updated to use rules_rust 0.16.1._
```sh
mkdir inherit_test
cd inherit_test
# Init bazel workspace
cat << 'EOF' > WORKSPACE
workspace(name = "inherit_from_workspace")
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
http_archive(
    name = "rules_rust",
    sha256 = "aaaa4b9591a5dad8d8907ae2dbe6e0eb49e6314946ce4c7149241648e56a1277",
    urls = ["https://github.com/bazelbuild/rules_rust/releases/download/0.16.1/rules_rust-v0.16.1.tar.gz"],
)
load("@rules_rust//rust:repositories.bzl", "rules_rust_dependencies", "rust_register_toolchains")
rules_rust_dependencies()
rust_register_toolchains(
    edition = "2021",
    versions = ["1.66.1"],
)
load("@rules_rust//crate_universe:defs.bzl", "crates_repository")
crates_repository(
    name = "crate_index",
    cargo_lockfile = "//:Cargo.lock",
    lockfile = "//:Cargo.Bazel.lock",
    manifests = [
        "//:Cargo.toml",
        "//sub/hello:Cargo.toml"
    ],
    rust_version = "1.66.1",
)
load("@crate_index//:defs.bzl", "crate_repositories")
crate_repositories()
EOF
# Add dummy BUILD file
touch BUILD
# Add empty lock files
touch Cargo.lock Cargo.Bazel.lock
# Init cargo workspace
cargo init --name inherit-depends --edition 2021 .
rm -r src
cat << 'EOF' > Cargo.toml
[workspace]
members = ["sub/hello"]
[workspace.dependencies]
anyhow = "1.0.65"
EOF
# Add package which uses workspace inheritance.
mkdir -p sub/hello/src
cat << 'EOF' > sub/hello/Cargo.toml
[package]
name = "hello"
version = "0.1.0"
edition = "2021"
[dependencies]
anyhow = { workspace = true }
EOF
cat << 'EOF' > sub/hello/BUILD
load("@crate_index//:defs.bzl", "all_crate_deps")
load("@rules_rust//rust:defs.bzl", "rust_binary")
rust_binary(
    name = "hello",
    srcs = glob(["src/**/*.rs"]),
    deps = all_crate_deps(
        normal = True,
    ),
)
EOF
cat << 'EOF' > sub/hello/src/main.rs
use anyhow::Result;

fn main() -> Result<()> {
    println!("hello world");
    Ok(())
}
EOF
```

Attempting to repin the dependencies produces an error:

```sh
CARGO_BAZEL_REPIN=true bazel sync --only=crate_index
```

<details>
<summary>Error result</summary>

```
Starting local Bazel server and connecting to it...
WARNING: ignoring _JAVA_OPTIONS in environment.
INFO: Repository crate_index instantiated at:
  /home/redacted/tmp/inherit_test/WORKSPACE:15:18: in <toplevel>
Repository rule crates_repository defined at:
  /home/redacted/.cache/bazel/_bazel_redacted/7ad356b0bf1137d4fa1a1478184dc92a/external/rules_rust/crate_universe/private/crates_repository.bzl:106:36: in <toplevel>
INFO: repository @crate_index' used the following cache hits instead of downloading the corresponding file.
 * Hash '3b0e0fcbcbacc585927eeddb665861507e1d6223fcf0b1feaccf56738da2d4a1' for https://github.com/bazelbuild/rules_rust/releases/download/0.16.1/cargo-bazel-x86_64-unknown-linux-gnu
If the definition of 'repository @crate_index' was updated, verify that the hashes were also updated.
ERROR: An error occurred during the fetch of repository 'crate_index':
   Traceback (most recent call last):
	File "/home/redacted/.cache/bazel/_bazel_redacted/7ad356b0bf1137d4fa1a1478184dc92a/external/rules_rust/crate_universe/private/crates_repository.bzl", line 73, column 22, in _crates_repository_impl
		execute_generator(
	File "/home/redacted/.cache/bazel/_bazel_redacted/7ad356b0bf1137d4fa1a1478184dc92a/external/rules_rust/crate_universe/private/generate_utils.bzl", line 459, column 21, in execute_generator
		result = execute(
	File "/home/redacted/.cache/bazel/_bazel_redacted/7ad356b0bf1137d4fa1a1478184dc92a/external/rules_rust/crate_universe/private/common_utils.bzl", line 44, column 13, in execute
		fail(_EXECUTE_ERROR_MESSAGE.format(
Error in fail: Command [/home/redacted/.cache/bazel/_bazel_redacted/7ad356b0bf1137d4fa1a1478184dc92a/external/crate_index/cargo-bazel, "generate", "--cargo-lockfile", /home/redacted/tmp/inherit_test/Cargo.lock, "--config", /home/redacted/.cache/bazel/_bazel_redacted/7ad356b0bf1137d4fa1a1478184dc92a/external/crate_index/cargo-bazel.json, "--splicing-manifest", /home/redacted/.cache/bazel/_bazel_redacted/7ad356b0bf1137d4fa1a1478184dc92a/external/crate_index/splicing_manifest.json, "--repository-dir", /home/redacted/.cache/bazel/_bazel_redacted/7ad356b0bf1137d4fa1a1478184dc92a/external/crate_index, "--cargo", /home/redacted/.cache/bazel/_bazel_redacted/7ad356b0bf1137d4fa1a1478184dc92a/external/rust_linux_x86_64__x86_64-unknown-linux-gnu__stable_tools/bin/cargo, "--rustc", /home/redacted/.cache/bazel/_bazel_redacted/7ad356b0bf1137d4fa1a1478184dc92a/external/rust_linux_x86_64__x86_64-unknown-linux-gnu__stable_tools/bin/rustc, "--lockfile", /home/redacted/tmp/inherit_test/Cargo.Bazel.lock, "--repin", "--metadata", /home/redacted/.cache/bazel/_bazel_redacted/7ad356b0bf1137d4fa1a1478184dc92a/external/crate_index/splicing-output/metadata.json] failed with exit code 1.
STDOUT ------------------------------------------------------------------------

STDERR ------------------------------------------------------------------------
Error: Failed to generate context digest

Caused by:
    0: Failed to load manifest '/home/redacted/tmp/inherit_test/sub/hello/Cargo.toml'
    1: This manifest requires workspace inheritance, but `inherit_workspace` hasn't been called yet

ERROR: /home/redacted/tmp/inherit_test/WORKSPACE:15:18: fetching crates_repository rule //external:crate_index: Traceback (most recent call last):
	File "/home/redacted/.cache/bazel/_bazel_redacted/7ad356b0bf1137d4fa1a1478184dc92a/external/rules_rust/crate_universe/private/crates_repository.bzl", line 73, column 22, in _crates_repository_impl
		execute_generator(
	File "/home/redacted/.cache/bazel/_bazel_redacted/7ad356b0bf1137d4fa1a1478184dc92a/external/rules_rust/crate_universe/private/generate_utils.bzl", line 459, column 21, in execute_generator
		result = execute(
	File "/home/redacted/.cache/bazel/_bazel_redacted/7ad356b0bf1137d4fa1a1478184dc92a/external/rules_rust/crate_universe/private/common_utils.bzl", line 44, column 13, in execute
		fail(_EXECUTE_ERROR_MESSAGE.format(
Error in fail: Command [/home/redacted/.cache/bazel/_bazel_redacted/7ad356b0bf1137d4fa1a1478184dc92a/external/crate_index/cargo-bazel, "generate", "--cargo-lockfile", /home/redacted/tmp/inherit_test/Cargo.lock, "--config", /home/redacted/.cache/bazel/_bazel_redacted/7ad356b0bf1137d4fa1a1478184dc92a/external/crate_index/cargo-bazel.json, "--splicing-manifest", /home/redacted/.cache/bazel/_bazel_redacted/7ad356b0bf1137d4fa1a1478184dc92a/external/crate_index/splicing_manifest.json, "--repository-dir", /home/redacted/.cache/bazel/_bazel_redacted/7ad356b0bf1137d4fa1a1478184dc92a/external/crate_index, "--cargo", /home/redacted/.cache/bazel/_bazel_redacted/7ad356b0bf1137d4fa1a1478184dc92a/external/rust_linux_x86_64__x86_64-unknown-linux-gnu__stable_tools/bin/cargo, "--rustc", /home/redacted/.cache/bazel/_bazel_redacted/7ad356b0bf1137d4fa1a1478184dc92a/external/rust_linux_x86_64__x86_64-unknown-linux-gnu__stable_tools/bin/rustc, "--lockfile", /home/redacted/tmp/inherit_test/Cargo.Bazel.lock, "--repin", "--metadata", /home/redacted/.cache/bazel/_bazel_redacted/7ad356b0bf1137d4fa1a1478184dc92a/external/crate_index/splicing-output/metadata.json] failed with exit code 1.
STDOUT ------------------------------------------------------------------------

STDERR ------------------------------------------------------------------------
Error: Failed to generate context digest

Caused by:
    0: Failed to load manifest '/home/redacted/tmp/inherit_test/sub/hello/Cargo.toml'
    1: This manifest requires workspace inheritance, but `inherit_workspace` hasn't been called yet

ERROR: no such package '@crate_index//': Command [/home/redacted/.cache/bazel/_bazel_redacted/7ad356b0bf1137d4fa1a1478184dc92a/external/crate_index/cargo-bazel, "generate", "--cargo-lockfile", /home/redacted/tmp/inherit_test/Cargo.lock, "--config", /home/redacted/.cache/bazel/_bazel_redacted/7ad356b0bf1137d4fa1a1478184dc92a/external/crate_index/cargo-bazel.json, "--splicing-manifest", /home/redacted/.cache/bazel/_bazel_redacted/7ad356b0bf1137d4fa1a1478184dc92a/external/crate_index/splicing_manifest.json, "--repository-dir", /home/redacted/.cache/bazel/_bazel_redacted/7ad356b0bf1137d4fa1a1478184dc92a/external/crate_index, "--cargo", /home/redacted/.cache/bazel/_bazel_redacted/7ad356b0bf1137d4fa1a1478184dc92a/external/rust_linux_x86_64__x86_64-unknown-linux-gnu__stable_tools/bin/cargo, "--rustc", /home/redacted/.cache/bazel/_bazel_redacted/7ad356b0bf1137d4fa1a1478184dc92a/external/rust_linux_x86_64__x86_64-unknown-linux-gnu__stable_tools/bin/rustc, "--lockfile", /home/redacted/tmp/inherit_test/Cargo.Bazel.lock, "--repin", "--metadata", /home/redacted/.cache/bazel/_bazel_redacted/7ad356b0bf1137d4fa1a1478184dc92a/external/crate_index/splicing-output/metadata.json] failed with exit code 1.
STDOUT ------------------------------------------------------------------------

STDERR ------------------------------------------------------------------------
Error: Failed to generate context digest

Caused by:
    0: Failed to load manifest '/home/redacted/tmp/inherit_test/sub/hello/Cargo.toml'
    1: This manifest requires workspace inheritance, but `inherit_workspace` hasn't been called yet

 checking cached actions
```
</details>

## Partial analysis
We observed that `crate_universe` appears to load the `Cargo.toml` file via [cargo_toml::Manifest::from_path](https://docs.rs/cargo_toml/0.13.0/cargo_toml/struct.Manifest.html#method.from_path) which in turn calls [cargo_toml::Manifest::complete_from_path](https://docs.rs/cargo_toml/0.13.0/cargo_toml/struct.Manifest.html#method.complete_from_path) for which the cargo_toml 0.13.0 documentation states:

> Note: this doesn’t support workspace inheritance yet. You need to get workspace’s manifest yourself, and then call inherit_workspace with it.

However, the documentation of the same method in [cargo_toml 0.14.0](https://docs.rs/cargo_toml/0.14.0/cargo_toml/struct.Manifest.html#method.complete_from_path) has changed to read:

> It supports workspace inheritance and will search for a root workspace. Use complete_from_path_and_workspace to provide the workspace explicitly.

## Proposed solution
So it looks like a simple dependency upgrade in crate_universe to cargo_toml `0.14.0` might resolve this issue.

**Comment by UebelAndre** [comment]
Thanks for reporting this! Would you be able to test what happens when `cargo_toml` is upgraded to `0.14.0`?

**Comment by joell** [comment]
I was able to build a version of the `cargo-bazel` binary and execute it directly using the arguments the Bazel error emitted.  With `cargo_toml` `0.13` it generated the same error, but with `0.14` the error went away and the command exited successfully.

I was not able to verify how that would work within the `rules_rust` Bazel rules, because I didn't find any developer documentation in this project, and it's not clear to me how to build and integrate it with a modified `cargo-raze` binary.

Hopefully, this limited confirmation will help, though.

**Comment by UebelAndre** [comment]
There are integration tests that will build `cargo-bazel`. There’s likely no test coverage for this feature though so that would need to be changed.

also, `cargo-raze` is a different project and has no integrations with `crate_universe`/`cargo-bazel`.

all you should need to do is update the Cargo.toml files, re-run the crate universe crates_vendor target, and write a regression test.

**Comment by joell** [comment]
Sorry.  `cargo-raze` was a typo.  (I used that previously, and finger memory betrayed me.)
I built and tested the `cargo-bazel` application in this repo's `crate_universe` directory.

The `cargo-bazel` binary I built failed with `cargo_toml` `0.13` and succeeded with `0.14`.

**Comment by UebelAndre** [comment]
> Sorry. `cargo-raze` was a typo. (I used that previously, and finger memory betrayed me.)

Ah, all good!

i guess, let’s start from the top and let me help you get this fix in. What information specifically do you need and once we’ve figured all that out, what would have helped you be more autonomous for this change.

**Comment by joell** [comment]
I finally found the `crate_universe/DEVELOPMENT.md` document which gave me just enough information to test my patched build of `cargo-bazel` with my full Bazel issue example.  It does in fact resolve the issue.  I've made pull request #1774 with the version-bump change.

I tried to look at how a test could be added that would have caught this problem, but in the limited time I had to look through the `cargo-bazel` test code, it was not clear where the test files were stored and coming from.  An example of adding a test for `cargo-bazel` would be an excellent addition to the `crate_universe/DEVELOPMENT.md` file.  (The repinning step that I missed in my pull request might be another nice thing to add to that document.)

Also, the first place I looked after the repo's front-page `README.md` was `CONTRIBUTING.md`, but it didn't give any instructions on how to do test builds.  Even a pointer to `crate_universe/DEVELOPMENT.md` and similar documents for other tools this repo depends on would have helped.
